### PR TITLE
Add CODE_OF_CONDUCT.md

### PR DESCRIPTION
**Pull request by behlendorf**
### Description

Team, we are putting together a code of conduct for OpenZFS.

The OpenZFS Code of Conduct will apply to spaces associated with the OpenZFS, ZFSonLinux, ZFSonOSX, and ZFSonWindows projects.

The OpenZFS community values respectful, welcoming behavior towards everyone. This enables our members to thrive and contribute, and encourages new participants to join our community.

Although most experiences within the community are positive, we all have moments when we are unaware of our tone or are too busy to choose our words carefully. The goal of this code of conduct is to set the tone for the community, and provide a process to resolve conflicts when they occur.

The proposed code of conduct is available here: http://www.open-zfs.org/wiki/Code_of_Conduct

The Code of Conduct was put together by Matt Ahrens, Karyn Ritter, and Brian Behlendorf, with input from Jorgen Lundman, Allan Jude, Joshua Clulow, Robert Mustacchi, Dan McDonald, Tom Caputi, Kirk McKusick, and Tim Chase.

If you have comments or questions about the code of conduct, please let us know. There is a 1-week community comment period before the Code of Conduct takes effect, so please get in touch before April 29th. You can reply to this PR, on the mailing list, or email any of us directly.

https://openzfs.topicbox.com/groups/developer/T5dc11696c7b4bddb-Mcc32cf77ccddb6c939528ab9

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (a change to man pages or other documentation)

### CODE REVIEW

**Comment by kpande** [comment]
> When joking, make sure what you say is not an insult to some people. 

how do you know if a joke is insulting to someone or not? what committee decides this?

**Comment by kpande** [comment]
> Posting sexually explicit or violent material.

so just to be clear, non-explicit sexual material is okay? I guess if you're making a joke, as long as it doesn't offend anyone. that seems to be a narrow window of allowance.

**Comment by behlendorf** [comment]
Use your best judgment, but keep in mind you're making a joke in public around strangers.  It's easy to be misunderstood and everyone gets this wrong from time to time.  Our goal is to provide a welcoming community for everyone.

**Comment by kpande** [comment]
what about comments like [this one](https://github.com/zfsonlinux/zfs/issues/8529#issuecomment-475900694) where the user has an offensive profile picture and demeanor toward project contributors?

if I recall correctly, nothing happened to *that* user, but I was chastised for how I suggested they go elsewhere to receive better help. which brings me to the question of what consequences are to be used when people break these seemingly arbitrary rules?

so if the guidance is "use your best judgement" how is this any different than what has been happening already?

**Comment by harrim4n** [comment]
Like @kpande I'm also "missing" a part in the CoC about who makes final decisions and how this process would look like.

I hope it never happens, but let's assume that two (or more) members of the community absolutely cannot get along for one reason or another and one (or both) parties violate the CoC. Who decides what the appropriate measures should be and how these measures look like and how they are enforced?

> While the community does have the recourse to remove those who do not comply with the code of conduct[...]

How would this look like, how would this "removal" be implemented and enforced? How does "the community" arrive at that conclusion?

Also, it is not clear who stands behind "conduct@open-zfs.org".

**Comment by kpande** [comment]
yes, what if one of the members with commit access is accused of "unwanted sexual contact"? do we allow the courts to rule on the case first? do we require a police report at all? does any kind of accusation result in negative consequences for the accused? is there a presumption of innocence?

and which consequences are they, exactly? are you going to reject PRs or other code contributions from the member? the Linux kernel has a code of conduct with similar wording, and high level developers were accused of heinous acts, and nothing happened, because the developer is so important that removing their commit access is not reasonable.

**Comment by harrim4n** [comment]
Re: How decisions are made:
This appears to be (at least partially) be documented here:
http://www.open-zfs.org/wiki/Working_Group_and_Code_of_Conduct_Mechanics

**Comment by richardelling** [comment]
[humor below]

> how do you know if a joke is insulting to someone or not? what committee decides this?

A committee of artists! Cue the Oscar Wilde sketch... http://www.montypython.net/scripts/oscar.php

**Comment by kpande** [comment]
so [changes are made](http://www.open-zfs.org/w/index.php?title=Code_of_Conduct&type=revision&diff=2677&oldid=2676) to the Code of Conduct without notifying us or having a "vote"?

**Comment by ahrens** [comment]
@harrim4n 
> Also, it is not clear who stands behind "conduct@open-zfs.org".

It's the "Code of Conduct Working Group":

http://www.open-zfs.org/wiki/Working_Group_and_Code_of_Conduct_Mechanics#Current_Working_Group_Members

The current Working Group members (last updated April 2019) are:

Matt Ahrens, matt@mahrens.org
Brian Behlendorf, behlendorf1@llnl.gov
Karyn Ritter, karyn.ritter@gmail.com

**Comment by ahrens** [comment]
@harrim4n 
> I'm also "missing" a part in the CoC about who makes final decisions and how this process would look like.

The process is pretty lightweight, but what we have is documented here:

http://www.open-zfs.org/wiki/Reporting_Guide#What_happens_after_you_file_a_report.3F

What happens after you file a report?

You will receive an email from the OpenZFS Code of Conduct Working Group acknowledging receipt immediately. The current Working Group members are:

Matt Ahrens, matt@mahrens.org
Brian Behlendorf, behlendorf@llnl.gov
Karyn Ritter, karyn.ritter@gmail.com
The working group will immediately meet to review the incident and determine:

What happened based on the perspective of all involved.
Whether this event constitutes a code of conduct violation.
Who the bad actor was.
Whether this is an ongoing situation, or if there is a threat to anyone's physical safety.
We promise to acknowledge receipt within 48 hours (and will aim for a faster response). If this is determined to be an ongoing incident or a threat to physical safety, the working groups' immediate priority will be to protect everyone involved. This means we may delay an "official" response until we believe that the situation has ended and that everyone is physically safe.

Once the working group has a complete account of the events, they will make a decision as to how to respond. Responses may include:

Nothing (if we determine no violation occurred).
A private gentle reminder of the code of conduct.
A public gentle reminder of the code of conduct.
A private reprimand from the working group to the individual(s) involved.
A public reprimand.
An imposed vacation (i.e. asking someone to "take a week off" from a mailing list, Slack, or IRC).
A permanent or temporary ban from some or all OpenZFS spaces (mailing lists, Slack, IRC, events, etc.)
We'll respond within one week to the person who filed the report with either a resolution or an explanation of why the situation is not yet resolved.

Once we've determined our final action, we'll contact the original reporter to let them know what action (if any) we'll be taking. We'll take into account feedback from the reporter on the appropriateness of our response, but we don't guarantee we'll act on it.

**Comment by ahrens** [comment]
@harrim4n 
> If there is an even number of members in the working group/committee and the "vote" (or whichever mechanism is used; it's not really documented) is split 50/50, what happens?

This process is intentionally left flexible for now.  My opinion is that if there's disagreement, the default should be towards taking less action.

**Comment by harrim4n** [comment]
@ahrens Thanks for the replies.
Maybe some more general feedback/my opinion:
I don't really see the problem that the introduction of this/any CoC solves.
As can be seen by @behlendorf 's reply "use your best judgement", the process won't and can't be objective anyways. So this/any CoC only introduces a framework to possibly **permanently** removing someone from the project without really improving the objectivity of any decisions that are made.
It also has some pretty serious potential for abuse, especially considering the "the committee appoints new members after taking the communities wishes into consideration" bit. Not trying to paint things blacker than they are, but project "takeovers" have happened in the past.

So, re: "leaving things flexible intentionally". IMO, the introduction of a CoC should be done very carefully and possible scenarios for abuse should be carefully considered.
Personally, my vote would be for no-CoC, as I don't see the problem that it solves, but you probably have more insight into this than I do.

**Comment by gmelikov** [comment]
I want to add my humble opinion on this case - I don't like overall idea of CoC, where we'll have something more than "just use common sense and moral standards" (just to have one to close this question, to be honest). You can't create a 100% specific script, so what do we want?

How often do we have offensive situations? IMHO not on a daily basis. I'd say they are very rare. Don't feed troll, and it dies.

**Comment by ikozhukhov** [comment]
> I want to add my humble opinion on this case - I don't like overall idea of CoC, where we'll have something more than "just use common sense and moral standards" (just to have one to close this question, to be honest). You can't create a 100% specific script, so what do we want?
> 
> How often do we have offensive situations? IMHO not on a daily basis. I'd say they are very rare. Don't feed troll, and it dies.

I agree with @gmelikov 
if you want to be an advocate of situation where you have to identify issues with different opinions, you have to learn culture of conflicted persons and try to investigate language barriers if they are try translate things from different languages to English and have mistakes in translation.

if you want try to resolve some conflicts of situations, you have to grow your room with persons from different countries and cultures - only native person can try to help to identify some language translation issues.

but current advocates room based only on English native speakers, and I'm sorry, it can be not enough in conflicts situations.

if you have rules - they are should be confirmed/to be agreements with others community members and based on votes, instead of just doc in repo like dictator ...

we need rules, but more flexible will be try to ask votes for agreements.

it's my opinion.

**Comment by ahrens** [comment]
@harrim4n Thanks for highlighting the word `permanently`.  Any decisions are subject to later re-evaluation, so it is more accurate to say that the most extreme action we would take is `indefinite`.  We'll update the document to reflect that.

As a reminder, any sort of removal of a person from the project would be undertaken only after all other options had been exhausted.

**Comment by behlendorf** [comment]
> How often do we have offensive situations? IMHO not on a daily basis. I'd say they are very rare.

I completely agree.  We're lucky to have a vibrant community where bad behavior is the exception and not the norm.  I personally expect that to continue to be the case.  Adding a code of conduct is one way we can better articulate these values.  I do not anticipate anything fundamentally changing.  The project is still managed by the same people with the same motivations and goals.

**Comment by kpande** [comment]
the fact that @behlendorf and @mahrens can't even respond to my comments in this thread gives me worry about them being 67% of the "Working Group". there have been multiple responses to _other_ comments in this thread.

**Comment by kpande** [comment]
> Adding a code of conduct is one way we can better articulate these values. I do not anticipate anything fundamentally changing. The project is still managed by the same people with the same motivations and goals.

if adding this does not change anything, then we should use mahrens' own words an 'err on the side of taking less action'.

if it is the same people managing the project, then we will have all the same issues as before.

**Comment by ahrens** [comment]
@kpande 
> so changes are made to the Code of Conduct without notifying us or having a "vote"?

We're changing the document in response to feedback during the comment period.  Next week, we'll post a summary of the changes made.

FYI, The Working Group will also review the Code of Conduct annually:

http://www.open-zfs.org/wiki/Working_Group_and_Code_of_Conduct_Mechanics#Revision_Policy_.2F_Process

One thing we can be sure of is that the community needs will change over time, and the Code of Conduct must evolve to keep up with these changing needs and views. The Working Group will review the Code of Conduct:

- on an annual basis, and make updates as needed based on the latest thinking for community behavior and
- as the need arises, the Working Group will make changes based on new information and community needs.

As previously mentioned, all changes will be sent to the community for a 1-week comment period before being committed to the public text.

**Comment by kpande** [comment]
> As previously mentioned, all changes will be sent to the community for a 1-week comment period before being committed to the public text.

what use is a comment period when the comments are ignored? a majority of the responses here indicate this is unnecessary. will those comments be taken into account? what does "taking that into account" look like? to me, it looks like they're being ignored - and these are project contributors' opinions we are talking about... the very people you are aiming to control.

so a 1 week comment period can just pass us by where 95% of responses indicate disagreement and then... what? where is the documented procedure for when people disagree with the working group?

**Comment by quantumpotato** [comment]
Long have I related ZFS 1.0.
This PR -- so close to 1.0 --greatly disturbs me for reasons I will outline here. I thank you for your time and welcoming me, a stranger, to your this discussion.

1st:

I do not understand the problem it is solving. All commits should solve a problem. The purpose of ZFS is to create a robust File System. All of its development has continued without a Code of Contact. Why now, does it need one?

2nd: 

This could lead to affirmative action.

```
Some individuals face more systemic challenges to their participation than others, and we strive to create a space that encourages participation regardless of gender identity and expression, sexual orientation, disabilities, neurodiversity, physical appearance, body size, ethnicity, nationality, race, age, religion, and similar personal characteristics.
```

As a transracial autistic veggiekin, I fear others like me being granted commit access to the ZFS repo on the basis of our disabilities or identity, because we are not, as a people, suited towards file system code. There are other, more suitable populations which could and should write ZFS code, like the fine maintainers of this project. Someone like me should only be granted access on the basis of code, not identity.

3rd: Null pointer exception: "Diversity" is undefined

```
Diversity is one of our huge strengths, but it can also lead to communication issues and unhappiness.
```

What exactly is diverse? If this is related to Point 2nd, please drop it. Who you are or what you believe doesn't matter when it comes To Writing File System Code. All that matters is that you Write File System Code.

4th: Unintended Consequences

```
Discriminatory language in any form.
```

I fear you are not aware of the Official Definition of Discriminatory, which would lead you to violating your own Code of Conduct accidentally.

```
discriminated; discriminating
Definition of discriminate
transitive verb

1a : to mark or perceive the distinguishing or peculiar features of
Depth perception may be defined as the ability to appreciate or discriminate the third dimension …
— H. G. Armstrong
b : DISTINGUISH, DIFFERENTIATE
discriminate hundreds of colors
2 : to distinguish by discerning or exposing differences : to recognize or identify as separate and distinct
discriminate right from wrong
especially : to distinguish from another like object
discriminate the individual voices in the choir
```

Now then, to say "This architecture is superior to that architecture" is a discriminatory statement.
To say "Not checking for null pointers is a worse habit than preventing null pointer exceptions" is a discriminatory statement.

_What exactly did you mean by this_?

**Comment by GregorKopka** [comment]
As this PR aims as adding a link toward a page that isn't under control of github.com/zfsonlinux this leads to carte blanche for the one(s) in control of the link target to change it without further notice to and particiation of github.com/zfsonlinux members. 
**This is at least questionable and should not be taken lightly.**

The file to be linked aims at regulating human behaviour using vague and unclear wording, effectively on the basis of a subjective morality that is freely to be defined by the members of the http://www.open-zfs.org/wiki/Working_Group_and_Code_of_Conduct_Mechanics. The mechanics outlined also give these members free hand to change all the rules to their liking, without any say of the ones being subjected to such a change (while the community might input comments/suggestions they're under no obligation to take them into account). While the current members might be the good guys who currently don't plan on turning evil... this can't be guaranteed for the future as it might not be the case for their successors. 
**Such power should not be granted to anyone or to any role.**

The referenced http://www.open-zfs.org/wiki/Code_of_Conduct_Summary sets the scope to also include `conduct outside these spaces`, which boils down to _everything one does_ being subjected of it.
**It is completely unreasonable to subject the private lives of others to 3rd party moralities.**

It also references http://www.open-zfs.org/wiki/Reporting_Guide as a mechanic for it, which among others sanctions what effectively is a pillory / public shaming under the guise of 'A public reprimand'. 
**This alone is absolutely despicable and supporting such ideas should be out of the question.**

Also so far no actual need for these kind of mechanics has been demonstrated (please post links in case I'm wrong on this) and as long as there is no need... there should be no need to install any mechanics, especially one like the proposed that can be weaponized that easily.

Bottom line: **This pull request should __not__ be merged.** 
Neither should any similar ones, should any pop up in case this one fails.

**Comment by pcd1193182** [comment]
@kpande 
> the fact that @behlendorf and @mahrens can't even respond to my comments in this thread gives me worry about them being 67% of the "Working Group". there have been multiple responses to _other_ comments in this thread.

You may have missed the part where Brian directly responded to 40% of your previous comments.

> > Adding a code of conduct is one way we can better articulate these values. I do not anticipate anything fundamentally changing. The project is still managed by the same people with the same motivations and goals.
> 
> if adding this does not change anything, then we should use mahrens' own words an 'err on the side of taking less action'.
> 
> if it is the same people managing the project, then we will have all the same issues as before.

1. Just because it won't make a fundamental change doesn't mean it will make no change.
2. Just because it won't make a fundamental change doesn't mean it isn't worth doing.

The changes made by the adoption of a CoC are largely felt in two ways. First, in the event of an issue that needs to be handled, the code of conduct provides guidelines for doing so and a process by which it can happen. That already isn't nothing; in fraught situations like the ones that tend to prompt code of conduct deliberations, having an established procedure can help stabilize things. Second, the mere presence of a code of conduct can help attract or retain developers and users, especially those who are more likely to be the subject of negative behaviors.


> > As previously mentioned, all changes will be sent to the community for a 1-week comment period before being committed to the public text.
> 
> what use is a comment period when the comments are ignored? a majority of the responses here indicate this is unnecessary. will those comments be taken into account? what does "taking that into account" look like? to me, it looks like they're being ignored - and these are project contributors' opinions we are talking about... the very people you are aiming to control.

The aim of the code of conduct is not to control, it is to protect and encourage. In the case where bad behavior happens, yes, some level of control is necessary to cause change, but it's certainly not the aim of the proposal.

> so a 1 week comment period can just pass us by where 95% of responses indicate disagreement and then... what? where is the documented procedure for when people disagree with the working group?

Just because 95% of the comments indicate disagreement doesn't mean that 95% of the community disagrees. Most people who don't feel strongly about something tend to not say much, and even among those who do feel strongly, those against a change or action are more likely to speak up, since the perceived status quo is the taking of the action in question (and people don't tend to do work unless they feel they have to).

@quantumpotato 
> 1st:
> 
> I do not understand the problem it is solving. All commits should solve a problem. The purpose of ZFS is to create a robust File System. All of its development has continued without a Code of Contact. Why now, does it need one?

Because a code of conduct helps to encourage our developers to behave decently towards each other and the users of the software. A code of conduct would *always* have been a good thing; just because we're only doing it now doesn't mean that it shouldn't be done. Encryption was only implemented for ZoL recently, but it's still a good idea.
 
> 2nd:
> 
> This could lead to affirmative action.
> 
> ```
> Some individuals face more systemic challenges to their participation than others, and we strive to create a space that encourages participation regardless of gender identity and expression, sexual orientation, disabilities, neurodiversity, physical appearance, body size, ethnicity, nationality, race, age, religion, and similar personal characteristics.
> ```
> 
> Code quality matters because this is a software project. Nothing else does.

I'm going to save myself some time and quote [myself in the email thread about the code of conduct](https://zfsonlinux.topicbox.com/groups/zfs-discuss/T957b6c8fe715fc7b-Md736de261aecef437c129952/zfs-discuss-openzfs-code-of-conduct) here:

"A project is not solely defined by its product; the people, time, and emotions that go into it are all important as well.  As someone who puts extensive time and energy into OpenZFS, I can think of plenty of achievements of value that were made possible by working in a supportive and nurturing environment, rather than a hostile and toxic one. If a CoC can help further that goal, then  it seems like an obvious win."

[and](https://zfsonlinux.topicbox.com/groups/zfs-discuss/T5cfd63ce013ea6f3-M30f9767d93c24ebfefaaf6d6/zfs-discuss-openzfs-code-of-conduct)

"I agree that number of contributors is important, but I also think that the diversity of those contributors is important. Studies have shown that more diverse groups make better decisions, and plenty of evidence has shown that monocultures breed failure. There  is also ample evidence in the form of software companies that seeking the myth of meritocracy above all else leads to a lack of diversity and a toxic culture."

> 
> 3rd: Null pointer exception: "Diversity" is undefined
> 
> ```
> Diversity is one of our huge strengths, but it can also lead to communication issues and unhappiness.
> ```
> 
> What exactly is diverse? If this is related to Point 2nd, please drop it. Who you are or what you believe doesn't matter when it comes To Writing File System Code. All that matters is that you Write File System Code.

Diversity is a fairly well-understood term in the common parlance, but usually it refers to the property of a group having a significantly heterogenous makeup across a variety of factors.  In this case, "gender identity and expression, sexual orientation, disabilities, neurodiversity, physical appearance, body size, ethnicity, nationality, race, age, religion, and similar personal characteristics" to quote the source you already quoted.

Also, there's plenty of things that matter besides writing code. Reviewing code, filing feature requests, testing, providing support, giving insight during design discussions, doing administrative tasks, keeping debates on topic, and many more things are all valuable contributions that can be made even by people who aren't systems programmers.

> 
> 4th: Unintended Consequences
> 
> ```
> Discriminatory language in any form.
> ```
> 
> I fear you are not aware of the Official Definition of Discriminatory, which would lead you to violating your own Code of Conduct accidentally.
> 
>  ...
> 
> Now then, to say "This architecture is superior to that architecture" is a discriminatory statement.
> To say "Not checking for null pointers is a worse habit than preventing null pointer exceptions" is a discriminatory statement.

Words have lots of meanings. That is one of the many problems with language generally, and english specifically. In this case, the meanings you highlighted are not the meanings that are intended by the document. The intended meanings are pretty clear, in my opinion, but in case they aren't to you:
```
In human social behavior, discrimination is treatment or consideration of, or making a distinction towards, a person based on the group, class, or category to which the person is perceived to belong. 
```
I think in light of that definition, this specific objection becomes void.

> 
> Subjective language is open to interpretation and misuse. Instead of trying to recreate a legal guidelines, I recommend operating the project as you have been, focusing on shipping quality software regardless of who is behind the commits. Some principles are best understood implicitly, without needing codification.

Thankfully, the people who wrote and designed this document, and selected it for inclusion, are still with us, and so can provide insight into what they meant and intended with their words. Also thankfully, the meanings of those words are pretty clear, especially given the context of the document. It requires a deliberate effort or very significant disconnect from the context of the words to misinterpret them in a way that would cause a problem, and such incidents are easily rectified by explaining the actual meaning of the words and phrases in use.

@GregorKopka 

> As this PR aims as adding a link toward a page that isn't under control of github.com/zfsonlinux this leads to carte blanche for the one(s) in control of the link target to change it without further notice to and particiation of github.com/zfsonlinux members.
> **This is at least questionable and should not be taken lightly.**

This is a procedural issue at most, and doesn't justify scrapping the PR. In addition, the document itself describes a revision process that requires a week of notice before making changes. That is certainly not carte blanche.


> The file to be linked aims at regulating human behaviour using vague and unclear wording, effectively on the basis of a subjective morality that is freely to be defined by the members of the http://www.open-zfs.org/wiki/Working_Group_and_Code_of_Conduct_Mechanics. The mechanics outlined also give these members free hand to change all the rules to their liking, without any say of the ones being subjected to such a change (while the community might input comments/suggestions they're under no obligation to take them into account). While the current members might be the good guys who currently don't plan on turning evil... this can't be guaranteed for the future as it might not be the case for their successors.
> **Such power should not be granted to anyone or to any role.**

So what, we should also not have courts of law or justice systems? The organizers of the community could already do evil things without a code of conduct in place if they didn't want to. This document doesn't create power from nothing, it merely clarifies what the powers are and when they should be used.

> The referenced http://www.open-zfs.org/wiki/Code_of_Conduct_Summary sets the scope to also include `conduct outside these spaces`, which boils down to _everything one does_ being subjected of it.
> **It is completely unreasonable to subject the private lives of others to 3rd party moralities.**

I will again cite myself to save time here:

"I'm fine with excluding people from the community who treat other members of the community badly. Just because someone says transphobic things at a bar after the ZFS developer summit instead of at it (to fabricate an example; any resemblance to real events is entirely unintended and accidental) doesn't mean they shouldn't be brought to task for their actions.  What's the alternative, exactly? We say it's fine to do whatever you want to community members as long as you do it outside of the few listed forums? That doesn't prevent harassment and discrimination, it just moves it to places we have less visibility into and makes people feel better about themselves ("look at us, we banned bad behavior!") without fixing anything."

> It also references http://www.open-zfs.org/wiki/Reporting_Guide as a mechanic for it, which among others sanctions what effectively is a pillory / public shaming under the guise of 'A public reprimand'.
> **This alone is absolutely despicable and supporting such ideas should be out of the question.**

Public shaming is not inherently evil. In moderation, it can correct or curb bad behavior.  Most justice  systems don't require a gag order on all punishments handed down, and it's not clear why this code should either. Abuses of public shaming can cause real problems, but the document you linked to several punishments both less and more severe than that one. If you are concerned about abuses of power, that point is already addressed

> Also so far no actual need for these kind of mechanics has been demonstrated (please post links in case I'm wrong on this) and as long as there is no need... there should be no need to install any mechanics, especially one like the proposed that can be weaponized that easily.

See my (numerous) previous responses regarding this point.

I approve of adding this code of conduct to the ZFS On Linux project.

**Comment by h1z1** [comment]
> "I'm fine with excluding people from the community who treat other members of the community badly. Just because someone says transphobic things at a bar after the ZFS developer summit

Shills are coming out in force.  Thankfully everything on Github is logged for enternity, I suspect this thread will soon be locked however.  If the first thing you're thinking is how to be offended by something someone says not only outside the context of anything remotely related to this project but at a bar, you've quite clearly demonstrated the problem with this.

The rest of your post is bable.

**Comment by kpande** [comment]
@pcd1193182  wrote:
>  The organizers of the community could already do evil things without a code of conduct in place

which brings an interesting clause in the CoC:
> Don’t attack the person or company that wrote the code. Focus on finding a solution to the problem.

and the behaviour of @ahrens censoring those who speak poorly of Oracle on Slack.

I guess once the CoC is in place, ahrens cannot blindly censor those who he disagrees with anymore? that'd be a welcome improvement.

**Comment by kpande** [comment]
> You may have missed the part where Brian directly responded to 40% of your previous comments.

why is responding to less than half of my concerns something you're here defending, when Brian is supposed to be one of the blessed and elite Working Group members, who is *going to have to listen to complaints*?

we get your opinion, Paul. you share Brian and Matt's opinion, which isn't surprising, considering one of them is paying you. please stop ramming your opinion down others' throats.

what we're asking for is for THE WORKING GROUP members to respond. thanks.

**Comment by ahrens** [comment]
> we get your opinion, Paul. you share Brian and Matt’s opinion, which isn’t surprising, considering one of them is paying you

Paul is employed by Delphix, not myself.  I am not Paul’s manager (I’m not anyone’s manager, I’m an individual contributor).  I didn’t ask Paul to contribute to this discussion.

I do agree with what Paul has said.

**Comment by ptx0** [comment]
you mean Delphix, the one company that has 2 of the 3 members on the Working Group?

**Comment by pcd1193182** [comment]
@h1z1 
> > "I'm fine with excluding people from the community who treat other members of the community badly. Just because someone says transphobic things at a bar after the ZFS developer summit
> 
> Shills are coming out in force. Thankfully everything on Github is logged for enternity, I suspect this thread will soon be locked. However if the first thing you're thinking is how to be offended by something someone says not only outside the context of anything remotely related to this project but at a bar, you've quite clearly demonstrated the problem with this.

Calling me a shill ("a person who publicly helps or gives credibility to a person or organization without disclosing that they have a close relationship with the person or organization.") is an interesting move. It's a little hard to support, given that I have multi-year history of involvement with the community, a large number of significant contributions (both commits and reviews) to my name, and have spoken at the OpenZFS Developer Summit. Certainly I make no secret of the fact that I work for Delphix, the company for which Matt also works, but I wasn't asked or instructed to make these posts.

My first thought was not that I was offended, it was "who's this person?" I'm still not offended, just sort of bemused.

> The rest of your post is bable.

As persuasive discourse goes, this lacks a certain something.

@kpande 
 
> which brings an interesting clause in the CoC:
> 
> > Don’t attack the person or company that wrote the code. Focus on finding a solution to the problem.
> 
> and the behaviour of @ahrens censoring those who speak poorly of Oracle on Slack.
> 
> I guess once the CoC is in place, ahrens cannot blindly censor those who he disagrees with anymore? that'd be a welcome improvement.

I haven't seen any instances of Matt censoring people on the OpenZFS slack, but I admit I don't follow it closely, and because our history is limited, I can't go back to look at the incident you're referring to. Do you happen to have more information about the incident? Certainly incidents of censorship would be concerning and worth discussing further in a public forum.


> > You may have missed the part where Brian directly responded to 40% of your previous comments.
> 
> why is responding to less than half of my concerns something you're here defending, when Brian is supposed to be one of the blessed and elite Working Group members, who is _going to have to listen to complaints_?

I'm not defending a 40% response rate, I'm pointing out your hyperbole.  In addition, there's a difference between comments on a Github thread and email chain and formally filed complaints. The former are much higher volume, and tend to have a worse signal:noise ratio.  As for why some of your comments weren't responded to, I couldn't say for sure, but my guess would be that

1. They're busy people and there have been a *lot* of comments in this thread and the email chain.
2. Some of your comments raise issues to which there isn't a clear answer.  It's hard to say what would be done in a hypothetical situation of the severity you describe, but I'm comfortable waiting until it happens and then letting the committee make a decision. If they make a decision I disagree with, I'll be happy to criticize them for it then.
3. Some of your comments have fairly clear answers. For example, your comment about "changes being made without a vote": Yes, changes can be made without a vote. The CoC makes that pretty clear. As for the lack of notice period, the document has not yet been finalized, so some changes are natural. The document describes that changes once the CoC is in effect, there will be another one-week comment period in the event of any changes.
4. Some of your comments involve dredging up past incidents that have been settled already. Your past conduct [has been addressed in the email thread.](https://zfsonlinux.topicbox.com/groups/zfs-discuss/T5cfd63ce013ea6f3-M56a3c7e5af184d7fca92e93e/zfs-discuss-openzfs-code-of-conduct)

> we get your opinion, Paul. you share Brian and Matt's opinion, which isn't surprising, considering one of them is paying you. please stop ramming your opinion down others' throats.

1.  Matt doesn't pay me. Matt isn't even my manager. Furthermore, Matt wasn't aware that I was going to contribute to this discussion until after I had already done so. I wasn't asked, instructed, directed, encouraged, or anything else to do so. This Github thread and the email chain weren't even *mentioned* at work (until yesterday morning) as far as I can recall. I'd prefer if you didn't accuse me of being a paid shill when I have done plenty of hard thinking about these issues and others related to them all on my own, especially because it's a pretty blatant ad hominem attack. Attack the arguments, not the person making them.
2. I'm confused, why should the fact that I disagree with you mean that I don't get to contribute to this discussion? You certainly seem to keep contributing despite the fact that your opinion has been made quite clear. There are no restrictions described anywhere as the maximum number of times anyone can post on this thread or the email chain.  I certainly am posting frequently (which I assume is what you're referring to as "ramming my opinion down others' throats"), but that's because I care deeply about this issue, because I have some experience with making reasoned arguments, and because I have the time and emotional labor to devote to doing so. None of those should be disqualifying; the opposite, in fact.

As a fun side note, attempting to silence people who disagree with you (e.g. by telling them to stop participating in a discussion) is a decent example of censorship. Your statement doesn't quite qualify, since you don't posses power to back it up (a key element of censorship), but it's certainly thematic.

> what we're asking for is for THE WORKING GROUP members to respond. thanks.

I can't speak for them, I can only speak for me. I suspect, though, that they might feel more compelled to post if there were serious concerns or arguments raised that hadn't already been addressed and/or rebutted.

**Comment by pcd1193182** [comment]
> you mean Delphix, the one company that has 2 of the 3 members on the Working Group?

1. Karyn is also not my manager (she's not even in the same part of the org chart), so if that's your concern, don't worry about it.
2.  Delphix as a whole really doesn't have a huge interest in the adoption of a Code of Conduct.  I think the totality of its effect on them is how it affects the future of the community and the project (which they do have a vested interest in, so you'd think if anything they'd be supporting moves that they thought would help it grow and thrive?), and how it affects their employees who are involved in that community (speaking personally, I have not been positively affected by this discussion. It has been draining, exhausting, and frustrating. But it's important, so I keep on keeping on).
3. This continues to be an ad hominem attack, which is a fallacy. Come on, guys. This is debate 101: even if I were holding a position I didn't fully believe (which, to reiterate, is *NOT* the case), the arguments are still valid until rebutted.

**Comment by ptx0** [comment]
> "I can't speak for them"

Okay then. let them answer.
